metalearners/slearner.py

```diff
@@ -40,28 +41,43 @@ def _append_treatment_to_covariates(
     X: Matrix, w: Vector, supports_categoricals: bool, n_variants: int
 ) -> Matrix:
     """Appends treatment columns to covariates and one-hot-encode if necessary."""
-    w = convert_treatment(w)
-    w = pd.Series(w, dtype="category").cat.set_categories(list(range(n_variants)))
+    w_np = convert_treatment(w)
 
-    if isinstance(X, pd.DataFrame) and "treatment" in X.columns:
+    # We enforce pandas as an intermediate data structure for the treatment vector w
+    # since we might rely on pandas' get_dummies function. We haven't found a narwhals
+    # alternative to get_dummies yet.
```

Comment:
We support [`Series.to_dummies`](https://narwhals-dev.github.io/narwhals/api-reference/series/#narwhals.series.Series.to_dummies) - you could first create a series from a numpy array and a given backend/native_namespace with [`nw.new_series`](https://narwhals-dev.github.io/narwhals/api-reference/narwhals/#narwhals.new_series) specifying `dtype=nw.Categorical()`.

However we don't support to `set_categories` 🤔

Reply:
Apologies on my end for the radio silence, I got drowned in other duties.

> From what I can see, polars categories and series.to_dummies would not really work for this use case due to the fact that you can have categories in list(range(n_variants)) but unseen in w_np.

I think you're absolutely right. I thought about whether we can get around that and I think for now we can't: we need to be able to handle the number of variants explicitly passed through `n_variants`.

I think my suggestion would be to pass through pandas for now and to watch your changelogs to see whether `polars` and `narwhals` might support `set_categories` at some point in the future. Does that sound reasonable to you @FBruzzesi ?

Reply:
As Narwhals is mostly a subset of the polars API, I may suggest to raise an issue in polars itself. I couldn't find any issue (closed or open) to require something like `set_categories` nor `to_dummy` to "respect" the list of categories when they do not appear in the series.

I would say you have a reasonable use case for those requests

Reply:
Sgtm! :)

https://github.com/pola-rs/polars/issues/21337